### PR TITLE
feat: allow users to pass offerArgs with their offer

### LINKED ISSUE #3574: Optional Args in E(zoe).offer(...
## What is the Problem Being Solved?

Currently, `E(zoe).offer` takes an invitation, a proposal, and the payments for an offer. This requires that contracts use the proposal as stating the entirety of the user's preferences. However, we have come across cases in which the user would really like to parameterize their offer more. 

The way that we've handled this, thus far, is that arguments can be passed into whatever function makes the invitation. However, the user may not be calling those methods, or even have access to them. There's also a temporal difference - making the invitation may be done much earlier than the use of the invitation.

## Description of the Design

We can change `E(zoe).offer` to take an `optionalArgs` object. This will be passed directly to the ZCFVat. The contract code can chose to validate and use or drop the object. 

## Security Considerations

Contract code that chooses to engage with the `optionalArgs` will need to protect themselves from anything the users might send. 

Additionally, we should ensure that Zoe cannot be affected by handling the `optionalArgs`. 

## Test Plan
A Swingset test where a user passes such an object in with their offer and a contract uses it.

### PR DESCRIPTION
**Pull request by katelynsills**
Changes `E(zoe).offer` to take an optional `offerArgs` object. The `offerArgs` are passed through Zoe to the offerHandler in the ZCFVat. The contract code can chose to validate and use or drop the object.

This gives the user a way to communicate preferences to the contract outside of the proposal. 

closes #3574

### CODE REVIEW

File: packages/zoe/src/contractFacet/zcfZygote.js
Comment (erights) [inline packages/zoe/src/contractFacet/zcfZygote.js:271]: This phrasing suggests that we enforce that reason is of type Error, rather than merely that it is undefined.

File: packages/zoe/src/zoeService/startInstance.js
Comment (erights) [inline packages/zoe/src/zoeService/startInstance.js]:
```suggestion
          offerArgs = undefined,
```

File: packages/zoe/src/internal-types.js
Comment (erights) [inline packages/zoe/src/internal-types.js:17]: I'm not sure if there's a proper way in TypeScript to distinguish between "this property may not exist" vs "this property's value may be undefined". In my ignorance of TypeScript, when I see {Object=} I leap to the first. But here you mean the second. How about
```suggestion
 * @property {Object | undefined} offerArgs
```

**Review by erights — COMMENTED** [review]
Review so far. Everything looks good but I'm holding off on LGTM approval until I try out one thing that may or may not work out. I should have an answer soon.

**Review by erights — APPROVED** [review]
See https://github.com/Agoric/agoric-sdk/pull/3584/files?diff=split&w=1

With that, LGTM.

**Comment by erights** [comment]
But note the "causes kernel panic" note in the PR comment at https://github.com/Agoric/agoric-sdk/pull/3584

**Comment by erights** [comment]
@warner @katelynsills Nevermind the kernel panic. After rebasing and `yarn build`ing, the problem went away. Sorry for the false alarm!

File: packages/zoe/src/contractFacet/zcfZygote.js
Comment (katelynsills) [reply packages/zoe/src/contractFacet/zcfZygote.js:271]: Hmm, I think we do want to enforce that the reason is of type Error. Making sure it's not undefined is part of that. Can we do better?

This code didn't change in this PR, so I think we keep this discussion going separately, after this PR gets merged.

File: packages/zoe/src/zoeService/startInstance.js
Comment (katelynsills) [reply packages/zoe/src/zoeService/startInstance.js]: Good point, I'll go fix any cases that don't make it clear that it may be undefined.

File: packages/zoe/src/internal-types.js
Comment (katelynsills) [reply packages/zoe/src/internal-types.js:17]: I think `Object=` is what is recommended for saying that a property may be undefined.

Screen Shot 2021-08-03 at 4 59 47 PM

Comment (katelynsills) [reply packages/zoe/src/internal-types.js:17]: Hmm, but specifying it explicitly as you do, means that the `?` is dropped:

Screen Shot 2021-08-03 at 5 01 29 PM




I wonder whether this has any meaningful difference, though, as doesn't nearly everything treat a property that doesn't exist and a property that exists but is undefined, the same?

Comment (katelynsills) [reply packages/zoe/src/internal-types.js:17]: I'm going to leave this as is for conformity with the rest of Zoe but we can continue this conversation.

**Pull request merged**
